### PR TITLE
Compiler: allow functions prototypes to have their own scopes

### PR DESCRIPTION
**Pull request by wrongnull**
Closes https://github.com/ziglang/zig/issues/15409
This patch adds scopes to functions prototypes. It may be useful for describing dependence of one parameter from another one. For instance:
```zig
const proto = fn (arg0: getType(), arg1: @TypeOf(arg0)) void;
```
With this capability compiler must be able to detect shadowing from current and enclosing scopes. However, when generics are involved in cases like the following:
```zig
const proto = fn(comptime T: type, arg: T)
```
compiler cannot infer second type and treats it as anytype. But the very same behavior appears in the latest compiler dev version with generic functions (not prototypes). I'm pretty sure that this doesn't apply to the original issue
Finally, it turned out that there was several local shadowings in the compiler_rt and std source code. I fixed it as well.

### CODE REVIEW

File: src/AstGen.zig [hunk 1]
```diff
@@ -1303,80 +1304,129 @@ fn fnProtoExpr(
...
+        }
+        const param_index: Zir.Inst.Index = @enumFromInt(astgen.instructions.len + 1);
+        _ = try align_gz.addBreakWithSrcNode(.break_inline, param_index, inst, fn_proto.ast.align_expr);
+        break :inst inst;
     };
 
     if (fn_proto.ast.addrspace_expr != 0) {
         return astgen.failNode(fn_proto.ast.addrspace_expr, "addrspace not allowed on function prototypes", .{});
     }
+    var addrspace_gz = block_scope.makeSubBlock(params_scope);
```
Comment (Vexu) [inline src/AstGen.zig:1372]: Unused.

File: src/AstGen.zig [hunk 2]
```diff
@@ -1303,80 +1304,129 @@ fn fnProtoExpr(
...
+    defer addrspace_gz.unstack();
 
     if (fn_proto.ast.section_expr != 0) {
         return astgen.failNode(fn_proto.ast.section_expr, "linksection not allowed on function prototypes", .{});
     }
+    var section_gz = block_scope.makeSubBlock(params_scope);
```
Comment (Vexu) [inline src/AstGen.zig:1378]: Also unused.

File: src/AstGen.zig [hunk 3]
```diff
@@ -10926,6 +10976,7 @@ const Scope = struct {
     /// The category of identifier. These tag names are user-visible in compile errors.
     const IdCat = enum {
         @"function parameter",
+        @"function prototype parameter",
```
Comment (Vexu) [inline src/AstGen.zig:10979]: I don't think there is any benefit to separating this from `@"function parameter"`.

File: src/AstGen.zig (hunk 1 above)
Comment (wrongnull) [reply src/AstGen.zig:1372]: > Unused.

It is but without this compiler crashes with 'use of null' in line 11328 in ```AstGen.zig``` in ```addFunc```. There is following doc comment in above this fuction 
```
/// Must be called with the following stack set up:
    ///  * gz (bottom)
    ///  * align_gz
    ///  * addrspace_gz
    ///  * section_gz
    ///  * cc_gz
    ///  * ret_gz
    ///  * body_gz (top)
```

File: src/AstGen.zig (hunk 2 above)
Comment (wrongnull) [reply src/AstGen.zig:1378]: > Also unused.

Same here

File: src/AstGen.zig (hunk 3 above)
Comment (wrongnull) [reply src/AstGen.zig:10979]: > I don't think there is any benefit to separating this from `@"function parameter"`.

I think with such wording compiler diagnostics are more accurate. But I can delete it if necessary

File: src/AstGen.zig (hunk 1 above)
Comment (Vexu) [reply src/AstGen.zig:1372]: Huh, it's strange how it properly checks each of them individually in the else branch but only `ret_gz` in the then branch.

**Comment by mlugg** [comment]
As mentioned on Discord, I do not think this is a desirable change. Allowing references to prior parameters when writing function _types_ implies stronger guarantees on generic function types than actually exist. For instance, under this change, the following equality holds:
```zig
fn (x: anytype) @TypeOf(x) == fn (x: anytype) ?@TypeOf(x)
```
Zig does not have a good way to represent specific generic function types at the type level; unless this changes. allowing you to write these types and then "demoting" them to the representable generic type (in this case written as `fn (anytype) anytype`) seems to me misleading at best.

File: src/AstGen.zig (hunk 1 above)
Comment (wrongnull) [reply src/AstGen.zig:1372]: Should I rewrite this check in addFn or let it be as it is?

**Comment by wrongnull** [comment]
> As mentioned on Discord, I do not think this is a desirable change. Allowing references to prior parameters when writing function _types_ implies stronger guarantees on generic function types than actually exist. For instance, under this change, the following equality holds:
> ```zig
> fn (x: anytype) @TypeOf(x) == fn (x: anytype) ?@TypeOf(x)
> ```
> Zig does not have a good way to represent specific generic function types at the type level; unless this changes. allowing you to write these types and then "demoting" them to the representable generic type (in this case printed as `fn (anytype) anytype`, although this is currently invalid syntax to actually write) seems to me misleading at best.

This behavior already in master. And users can face to the very same issue using ```@TypeOf``` . In this case this patch just preverves behavior that already exist.
```zig
fn foo(x: anytype) @TypeOf(x) {
    return x;
}

fn bar(x: anytype) ?@TypeOf(x) {
    return null;
}


pub fn main() void {
    @compileLog(@TypeOf(foo) == @TypeOf(bar));
}
```

```
Compile Log Output:
@as(bool, true)
```

**Comment by mlugg** [comment]
> This behavior already in master. [...] In this case this patch just preserves already existing behavior

No, this patch takes a behavior from *one* part of the language, and applies it to a *different* part of the language. Consistency between these syntax forms is not itself a fundamental end goal here: to make this change, you must justify why allowing backreferences is useful/important despite these equivalences. Justifying this is the job of a language proposal.

In function *declarations*, we have to allow backreferences, because this is fundamental to Zig's system of generics. The way the type system interacts with generics there is certainly unfortunate, but it does not call for us to eliminate backreferences entirely, since the type of a generic function is only a small detail of it (it's actually fairly rare to interact much, if at all, with generic function types in Zig code). On the other hand, in function *types*, there is _zero practical use_ for these backreferences in a way which is not explicitly misleading - writing generic function types, where (since the whole thing you're doing is writing a type) the syntax strongly implies that the type is encoding the constraints you are writing when, in reality, it is not.

I believe this change should be a language proposal (and it is a proposal I would personally oppose).

**Comment by andrewrk** [comment]
I've converted #15409 into a language proposal; we can continue the discussion there. This code is still available to be resurrected depending on the outcome.

**Pull request closed without merging**